server/core/region.go

```diff
@@ -893,6 +939,47 @@ func DiffRegionPeersInfo(origin *RegionInfo, other *RegionInfo) string {
 	return strings.Join(ret, ",")
 }
 
+// isEqualPeers return the peers is equal or not
+func isEqualPeers(origin []*metapb.Peer, other []*metapb.Peer) bool {
```

Comment:
Why need this? I think it's easy to ignore update this function if we add some items in this meta proto.

Reply:
The update will not be missed, because the store.update (region) method will be called in the addRegionToSubTree method, which will determine that this region exists in the tree, and will update this region:
```
func (rst *regionSubTree) update(region *RegionInfo) {
	if r := rst.find(region); r != nil {
		rst.totalSize += region.approximateSize - r.region.approximateSize
		rst.totalKeys += region.approximateKeys - r.region.approximateKeys
		r.region = region
		return
	}
```
The following code is mainly to prevent missed deletion, for example: when the leader transfer peer occurs, we must delete this region in the subtree and then add it to the subtree:
```
if origin := r.regions.Get(region.GetID()); origin != nil {
			remove := false
			if origin.leader.GetId() != region.leader.GetId() {
				remove = true
			} else if !isEqualPeers(origin.GetVoters(), region.GetVoters()) {
				remove = true
			} else if !isEqualPeers(origin.GetLearners(), region.GetLearners()) {
				remove = true
			} else if !isEqualPeers(origin.GetPendingPeers(), region.GetPendingPeers()) {
				remove = true
			}
			if remove {
				r.removeRegionFromSubTree(origin)
			}
		}
```

Reply:
Yes, I mean this logic not easy to maintain.  If we add a new option in `metapb.Peer`, how do you know should we update the funciton `isEualPeers`?

Reply:
I reply on slack.

Reply:
I agree that this part of the code is hard to maintain.